fe/fe-core/src/main/java/org/apache/doris/nereids/rules/implementation/LogicalWindowToPhysicalWindow.java

```diff
@@ -0,0 +1,539 @@
+                        partitionKeyGroupList.remove(pkg2);
+                        merged = true;
+                        break;
+                    }
+                }
+                if (merged) {
+                    break;
+                }
+            }
+        }
+    }
+
+    /**
+     * [This comment references AnalyticPlanner.orderGroups()]
+     *
+     * Order partition groups (and the SortGroups, WindowFrameGroups within them) by increasing
+     * totalOutputTupleSize. This minimizes the total volume of data that needs to be
+     * repartitioned and sorted.
+     *
+     * Always move the non-partitioning partition group to the end, if it exists.
+     */
+    private void sortPartitionKeyGroups(List<PartitionKeyGroup> partitionKeyGroupList) {
+
+        PartitionKeyGroup noPartition = null;
+        // after createCommonPartitionKeyGroups(), there will be at most one empty partition.
+        for (PartitionKeyGroup pkg : partitionKeyGroupList) {
+            if (pkg.partitionKeyList.isEmpty()) {
+                noPartition = pkg;
+                partitionKeyGroupList.remove(noPartition);
+                break;
+            }
+        }
+
+        if (noPartition != null) {
+            partitionKeyGroupList.remove(noPartition);
+        }
+
+        partitionKeyGroupList.sort((pkg1, pkg2) ->
+                Integer.compare(pkg1.getTupleSize() - pkg2.getTupleSize(), 0)
+        );
+
+        if (noPartition != null) {
+            partitionKeyGroupList.add(noPartition);
+        }
+
+        for (PartitionKeyGroup pkg : partitionKeyGroupList) {
+            sortOrderKeyGroups(pkg.getGroupList());
+        }
+
+    }
+
+    private void sortOrderKeyGroups(List<OrderKeyGroup> orderKeyGroupList) {
+        orderKeyGroupList.sort((okg1, okg2) ->
+                Integer.compare(okg1.getTupleSize() - okg2.getTupleSize(), 0)
+        );
+
+        for (OrderKeyGroup okg : orderKeyGroupList) {
+            sortWindowFrameGroups(okg.getGroupList());
+        }
+    }
+
+    private void sortWindowFrameGroups(List<WindowFrameGroup> windowFrameGroupList) {
+        windowFrameGroupList.sort((wfg1, wfg2) ->
+                Integer.compare(wfg1.getTupleSize() - wfg2.getTupleSize(), 0)
+        );
+    }
+
+    /**
+     * Window Functions that have common PartitionKeys, OrderKeys and WindowFrame
+     */
+    public static class WindowFrameGroup extends WindowFunctionRelatedGroup<NamedExpression> {
+
+        private final List<Expression> partitionKeyList;
+        private final List<OrderKey> orderKeyList;
+        private final WindowFrame windowFrame;
+
+        public WindowFrameGroup(NamedExpression windowAlias) {
+            Window window = (Window) (windowAlias.child(0));
+            partitionKeyList = window.getPartitionKeyList().orElse(Lists.newArrayList());
+            orderKeyList = window.getOrderKeyList().orElse(Lists.newArrayList());
+            windowFrame = window.getWindowFrame().orElse(null);
```

Comment:
orElse should return default windowFrame
RANGE BETWEEN UNBOUNDED_PRECEDING AND CURRENT_ROW